lib/std/os/linux/io_uring.zig

```diff
@@ -0,0 +1,877 @@
+    /// Returns a pointer to the SQE so that you can further modify the SQE for advanced use cases.
+    /// A no-op is more useful than may appear at first glance.
+    /// For example, you could call `drain_previous_sqes()` on the returned SQE, to use the no-op to
+    /// know when the ring is idle before acting on a kill signal.
+    pub fn queue_nop(self: *IO_Uring, user_data: u64) !*io_uring_sqe {
+        const sqe = try self.get_sqe();
+        sqe.opcode = .NOP;
+        sqe.user_data = user_data;
+        return sqe;
+    }
+
+    /// Queues (but does not submit) an SQE to perform a `read(2)`.
+    /// Returns a pointer to the SQE.
+    pub fn queue_read(
+        self: *IO_Uring,
+        user_data: u64,
+        fd: os.fd_t,
+        buffer: []u8,
+        offset: u64
+    ) !*io_uring_sqe {
+        const sqe = try self.get_sqe();
+        sqe.opcode = .READ;
+        sqe.fd = fd;
+        sqe.off = offset;
+        sqe.addr = @ptrToInt(buffer.ptr);
+        sqe.len = @truncate(u32, buffer.len);
+        sqe.user_data = user_data;
+        return sqe;
+    }
+
+    /// Queues (but does not submit) an SQE to perform a `write(2)`.
+    /// Returns a pointer to the SQE.
+    pub fn queue_write(
+        self: *IO_Uring,
+        user_data: u64,
+        fd: os.fd_t,
+        buffer: []const u8,
+        offset: u64
+    ) !*io_uring_sqe {
+        const sqe = try self.get_sqe();
+        sqe.opcode = .WRITE;
+        sqe.fd = fd;
+        sqe.off = offset;
+        sqe.addr = @ptrToInt(buffer.ptr);
+        sqe.len = @truncate(u32, buffer.len);
+        sqe.user_data = user_data;
+        return sqe;
+    }
+
+    /// Queues (but does not submit) an SQE to perform a `preadv()`.
+    /// Returns a pointer to the SQE so that you can further modify the SQE for advanced use cases.
+    /// For example, if you want to do a `preadv2()` then set `rw_flags` on the returned SQE.
+    /// See https://linux.die.net/man/2/preadv.
+    pub fn queue_readv(
+        self: *IO_Uring,
+        user_data: u64,
+        fd: os.fd_t,
+        iovecs: []const os.iovec,
+        offset: u64
+    ) !*io_uring_sqe {
+        const sqe = try self.get_sqe();
+        sqe.opcode = .READV;
+        sqe.fd = fd;
+        sqe.off = offset;
+        sqe.addr = @ptrToInt(iovecs.ptr);
+        sqe.len = @truncate(u32, iovecs.len);
+        sqe.user_data = user_data;
+        return sqe;
+    }
+
+    /// Queues (but does not submit) an SQE to perform a `pwritev()`.
+    /// Returns a pointer to the SQE so that you can further modify the SQE for advanced use cases.
+    /// For example, if you want to do a `pwritev2()` then set `rw_flags` on the returned SQE.
+    /// See https://linux.die.net/man/2/pwritev.
+    pub fn queue_writev(
+        self: *IO_Uring,
+        user_data: u64,
+        fd: os.fd_t,
+        iovecs: []const os.iovec_const,
+        offset: u64
+    ) !*io_uring_sqe {
+        const sqe = try self.get_sqe();
+        sqe.opcode = .WRITEV;
```

Comment:
Me too, and that's what we had, but then the `io_uring_sqe` struct must have default 0 values for all fields except `opcode`, because there's going to be many opcodes at the end of the day, and our queue methods will otherwise become boilerplate. I do think default 0 values makes sense for this struct.

Reply:
If we switch back to default 0 values and the initialization pattern, then we can drop the `@memset` in `get_sqe()`.

Reply:
I'm willing to make the change back again, and think we should, but would like @andrewrk to weigh in.

Reply:
@daurnimator, as requested, we're now using explicit struct initialization again, without default zero values. To reduce boilerplate, we now have the `io_uring_prep_rw()` etc. methods straight from liburing. Having these `io_uring_prep_` methods decoupled from the queueing methods also allows the advanced non-sequential SQE allocation schemes you were thinking about (reusing static SQE slots etc. via array indirection), while still enabling the more common usage of sequential SQE allocation provided by `get_sqe()` from liburing. I think this addresses all concerns and improves the interface significantly.